src/tunnel.cc

```diff
@@ -391,12 +398,20 @@ TunnelStateData::checkRetry()
 {
     if (shutting_down)
         return "shutting down";
+    if (exhaustedTries())
+        return "exhausted tries";
     if (!FwdState::EnoughTimeToReForward(startTime))
         return "forwarding timeout";
-    if (!retriable)
-        return "not retriable";
+    if (banRetries)
+        return banRetries;
     if (noConnections())
         return "no connections";
+
+    // TODO: Use Optional for peer_reply_status to avoid treating zero value specially.
+    if (request->hier.peer_reply_status != Http::scNone && !Http::IsReforwardableStatus(request->hier.peer_reply_status))
+        return "received HTTP status code is not reforwardable";
+
+    // TODO: check pinned connections; see FwdState::pinnedCanRetry()
```

Comment:
Completing this TODO may change Squid behavior in splice-at-step2 deployments (at least). I think it is best done in a dedicated PR, especially since this PR is not really about SslBump.

Reply:
It looks incorrect to ```return nullptr``` in this case.

IMO this should forbid retry until that later PR implements the bump edge case(s) correctly.

Reply:
> It looks incorrect to `return nullptr` in this case.

I agree.

> IMO this should forbid retry until that later PR implements the bump edge case(s) correctly.

IMO this _change_ in the official Squid behavior (and a bug fix) deserve a dedicated PR for several reasons, including this one: IIRC, the last time we adjusted similar FwdState code, there was no consensus regarding (of even sufficient knowledge about!) how certain esoteric pinned cases should be handled. We should proceed with caution and isolate those changes in a dedicated PR.

Thank you for not insisting on doing these changes in this PR.

Reply:
> We should proceed with caution and isolate those changes in a dedicated PR.

FWIW, I have created the corresponding internal project and hope to see this aspect improved (and this TODO addressed) in the foreseeable future.